### ISSUE TITLE
hddtemp_smartctl wakes SAS disks

### ISSUE THREAD

**Issue by nickcmaynard** [munin-monitoring/munin#1484]
Hello!

Seagate SAS disks configured with hardware-based automatic standby logic (using Standby Condition Timers, etc., per [Seagate's whitepaper](https://www.seagate.com/files/docs/pdf/whitepaper/tp608-powerchoice-tech-provides-us.pdf)) are woken up by both `hdparm -C` and `smartctl -n standby`.

Indeed, the following power conditions are all cancelled: "STANDBY_Z", "STANDBY_Y", "IDLE_C", "IDLE_B", "IDLE".  This represents all of the low-power operation modes supported.

In these situations, an additional REQUEST SENSE check is required, using either `sdparm` or the `sg_request` part of the `sg3_utils` package:

```sh
# sdparm
nick@chester:/dev/disk/by-id$ sudo sdparm --readonly --command=sense /dev/sdk
    /dev/sdk: SEAGATE   ST6000NM0034      E005
Additional sense: Standby condition activated by command

# sg_requests
nick@chester:/dev/disk/by-id$ sudo sg_requests /dev/sdk
data-in decoded as sense:
Fixed format, current; Sense key: No Sense
Additional sense: Standby condition activated by command
```

The critical part here is that the "standby condition" has been activated.

This is also encoded as hex data, which is more easily parsed:

```sh
# sdparm
nick@chester:/dev/disk/by-id$ sudo sdparm --readonly --command=sense --hex /dev/sdk
 00     70 00 00 00 00 00 00 0a  00 00 00 00 5e 04 00 00
 10     00 00

# sg_requests
nick@chester:/dev/disk/by-id$ sudo sg_requests --hex /dev/sdk
 00     70 00 00 00 00 00 00 0a  00 00 00 00 5e 04 00 00
 10     00 00
```

Note the `5e`=`04` byte pairing, indicating the drive had been forced into standby (for testing purposes). The possible values are outlined in Seagate's [reference](https://www.seagate.com/files/docs/pdf/whitepaper/tp608-powerchoice-tech-provides-us.pdf), excerpt:

> Determining the Current Power State of the Drive
> To allow the host to determine whether the drive is in a power condition, a SCSI Request Sense command
> can be issued. Receipt of a Request Sense command does not change the current power state of the drive. If
> the drive is currently in a power condition, the reported sense code is as shown:

> | Sense Key | ASCQ | Description |
> | --- | --- | --- | 
> | 5Eh | 00h | LOW POWER CONDITION ON |
> | 5Eh | 01h | IDLE CONDITION ACTIVIATED BY TIMER
> | 5Eh | 02h | STANDBY CONDITION ACTIVIATED BY TIMER
> | 5Eh | 03h | IDLE CONDITION ACTIVIATED BY COMMAND
> | 5Eh | 04h | STANDBY CONDITION ACTIVIATED BY COMMAND
> | 5Eh | 05h | IDLE B CONDITION ACTIVIATED BY TIMER
> | 5Eh | 06h | IDLE B CONDITION ACTIVATED BY
> | 5Eh | 07h | IDLE C CONDITION ACTIVATED BY TIMER
> | 5Eh | 08h | IDLE C CONDITION ACTIVATED BY COMMAND
> | 5Eh | 09h | STANDBY Y CONDITION ACTIVATED BY TIMER
> | 5Eh | 0Ah | STANDBY Y CONDITION ACTIVATED BY

Indeed any value, and definitely non-zero values, indicate a low-power state of some description.  I have not seen a `5e` = `00` pairing in my testing - rather, if the drive is active, the `5e` sense key is not present at all.

Thus, this plugin should:
1. Interrogate `sdparm`, if present, for sense data
2. Parse the sense data, searching for sense key `5Eh`
3. (Optionally, see above) If found, check key for a non-zero hex value
4. Skip further efforts if found

**Comment by steveschnepp** [comment]
My naive opinion is that I don't think that it would be wise to make the plugin too clever, as it will be very hard to debug.

Nonetheless, I understand the need. Therefore, it would be better to have a (new) 2-stage plugin. One stage (daemon) to gather the informations, and another that is polled by the munin master. There, the daemon would do its measures whenever it can to avoid spinning the disks up. It could even record the "downtime".

**Comment by niclan** [comment]
I share @steveschnepp 's opinion: all that is way too much. If there isn't a simple option to avoid waking the SAS disks the operator should instead disable monitoring of them.

**Comment by nickcmaynard** [comment]
OK.  Apologies for all the context - I was just illustrating my thinking.  It may have made people think "eek, too hard!"

TL;DR: there is a very simple option to cover these (and other SAS) disks, outlined in the final 4 lines.

Basically, run sdparm, look at the hex result, and skip checking if there's a 5e sense key is present.  Sure, it's a little more involved than checking a status code, but it's pretty simple!

**Comment by nickcmaynard** [comment]
PR attached.
